### PR TITLE
Add net.launchpad.qpdfview

### PR DESCRIPTION
**Pull request by hg**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.

qpdfview is an excellent tabbed document reader that supports PDF, DjVu, and PostScript formats and is the closest analog to SumatraPDF that I could find. I've used it for years.

---

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] ~~I am an author/developer/upstream contributor of the project.~~ If not, I contacted upstream developers about this submission.

I wrote the manifest a little over a year ago, and [asked the upstream][q] to consider submitting it themselves, but have received no feedback.

[q]: https://answers.launchpad.net/qpdfview/+question/707222

---

## filesystem access

Because of `filesystem=host:ro`, users will not be able to save copies of documents from within the application, and this functionality is disabled at build time. Considering how exploitable PDF readers are, and how niche this feature is, I think we should rely on sandboxing Flatpak provides us to protect the majority of users from e.g. cryptominers.

Another option is to *not* disable the menu entries, and somehow let users who need this feature know that they should override the default permission.

What do you think?

## metainfo

I understand that metainfo changes should be upstreamed, but this project runs at slow pace, with releases being cut every 2-4 years. If we decide to block on this, the submission will be delayed by at least another year.

I will try to upstream these changes if/when this is merged, and (if successful) will remove the patch on next release.

Original screenshots have very low resolution, so I made my own.

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

**Comment by bbhtt** [comment]
Pretty sure `net.launchpad` ID means it would be unverifiable. Does launchpad allow project's to have static sites on a subdomain? What does the app internally use as ID?

File: metainfo.patch [hunk 1]
```diff
@@ -0,0 +1,50 @@
+diff --git a/miscellaneous/qpdfview.appdata.xml b/miscellaneous/qpdfview.appdata.xml
+index bdcc460..6ddce9a 100644
+--- a/miscellaneous/qpdfview.appdata.xml
++++ b/miscellaneous/qpdfview.appdata.xml
+@@ -32,24 +32,39 @@
+     <launchable type="desktop-id">qpdfview.desktop</launchable>
+     <screenshots>
+       <screenshot type="default">
+-        <image>https://screenshots.debian.net/shrine/screenshot/14328/simage/small-587677349ebef2f9ab1b56d7cbc7888e.png</image>
++        <image>https://i.imgur.com/VzApQWY.png</image>
```
Comment (bbhtt) [inline metainfo.patch:10]: I'd just commit them to the PR here under `screenshots/`.

File: net.launchpad.qpdfview.yml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+rename-appdata-file: qpdfview.appdata.xml
+
+finish-args:
+  - --filesystem=host:ro # open file dialog; drag & drop
+  - --device=dri # hardware acceleration
+  - --share=ipc # x11
+  - --socket=wayland # gui
+  - --socket=fallback-x11 # gui
+  - --socket=cups # printing
+  - --own-name=local.qpdfview # unique instance
```
Comment (bbhtt) [inline net.launchpad.qpdfview.yml:19]: Does the app really use `.local` ?

File: net.launchpad.qpdfview.yml [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+cleanup:
+  - /include
+  - /lib/*.a
+  - /lib/*.la
```
Comment (bbhtt) [inline net.launchpad.qpdfview.yml:24]:
```suggestion
  - "*.a"
  - "*.la"
```

File: net.launchpad.qpdfview.yml [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+          project-id: 1724
+          stable-only: true
+          url-template: http://libspectre.freedesktop.org/releases/libspectre-$version.tar.gz
+
+  - name: qpdfview
+    buildsystem: simple
+    build-commands:
+      # Saving copies will not be available because of `filesystem=host:ro`
+      - sed -i 's|return m_document->canSave();|return false;|' sources/documentview.cpp
+      - sed -i 's|/usr|/app|' qpdfview.pri
```
Comment (bbhtt) [inline net.launchpad.qpdfview.yml:136]: Can you use a patch for this?

**Comment by bbhtt** [comment]
bot, build net.launchpad.qpdfview

File: net.launchpad.qpdfview.yml [hunk 5]
```diff
@@ -0,0 +1,156 @@
+app-id: net.launchpad.qpdfview
+default-branch: stable
```
Comment (bbhtt) [inline net.launchpad.qpdfview.yml:2]:
```suggestion
```

File: net.launchpad.qpdfview.yml (hunk 2 above)
Comment (hfiguiere) [reply net.launchpad.qpdfview.yml:19]: this is a bug if that's the case.

File: net.launchpad.qpdfview.yml [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+          url-template: http://libspectre.freedesktop.org/releases/libspectre-$version.tar.gz
+
+  - name: qpdfview
+    buildsystem: simple
+    build-commands:
+      # Saving copies will not be available because of `filesystem=host:ro`
+      - sed -i 's|return m_document->canSave();|return false;|' sources/documentview.cpp
+      - sed -i 's|/usr|/app|' qpdfview.pri
+      - lrelease qpdfview.pro
+      - qmake qpdfview.pro
```
Comment (hfiguiere) [inline net.launchpad.qpdfview.yml:138]: should this be `buildsystem: qmake` ?

File: metainfo.patch (hunk 1 above)
Comment (bbhtt) [reply metainfo.patch]: Use these links for now https://raw.githubusercontent.com/flathub/flathub/8f5611acd0a2a6cc4a58c130ae8d952f5af9b226/screenshots/1-main.png

You can update it once the submission is accepted into a new repository.

File: net.launchpad.qpdfview.yml (hunk 2 above)
Comment (bbhtt) [reply net.launchpad.qpdfview.yml:18]: This should be fixed upstream. You should probably open an issue

**Comment by bbhtt** [comment]
bot, build net.launchpad.qpdfview

**Comment by bbhtt** [comment]
I'm a bit concerned about the ID, as the developers of this software don't own, `launchpad.net`, which is under the control of Canonical. If upstream is active, the issue about correct dbus names should be raised.

The ID can follow the proper dbus name if that is fixed.

**Pull request closed without merging**